### PR TITLE
article: hierarchical config - interim solution

### PR DESCRIPTION
**Pull request by rawkode**

### CODE REVIEW

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+When I say "really common" ... I mean really really common.
+
+This [issue](https://github.com/pulumi/pulumi/issues/2307) was first open in 2018 and has received 52 upvotes from the community. Not only that, we've had plenty of similiar issues created over the years too.
+
+- [Feature-Request: project-wide secrets #2445](https://github.com/pulumi/pulumi/issues/2445)
+- [Feature Request: Global Config Values](https://github.com/pulumi/pulumi-aws/issues/1052)
+- [How to share a config between projects](https://github.com/pulumi/pulumi/issues/5473)
+- [Project-wide variables (not stack specific) #6719](https://github.com/pulumi/pulumi/issues/6719)
+
+This is clearly a feature that our community is asked for and we're currently working on delivering it as soon as we can. However, in the interim, I'm going to show you a few ways that you can get around this now.
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:22]:
```suggestion
This is clearly a feature that our community has asked for and we're currently working on delivering it as soon as we can. However, in the interim, I'm going to show you a few ways that you can get around this now.
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+draft: false
+meta_desc: In this article, we deploy k3s and use the Command package to retrieve our kubeconfig from the virtual-machine and create a Kubernetes provider
+meta_image: meta.png
+authors: ["david-flanagan"]
+tags: ["Config"]
+---
+
+A really common question that we receive on the Pulumi team is, "How can we set config at a project level, that can be used across all stacks?".
+
+When I say "really common" ... I mean really really common.
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:13]: Nit: 

```suggestion
When I say "really common" ... I mean really, really common.
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+conf := config.New(ctx, "")
+region := conf.Require("region")
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+```python
+config = pulumi.Config();
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:57]: We show the imports in the TS and Go examples -- can we show them here as well?

```suggestion
import pulumi

config = pulumi.Config();
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+region = config.require('region');
+```
+
+{{% /choosable %}}
+
+{{% choosable language csharp %}}
+
+```csharp
+var config = new Pulumi.Config();
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:66]: Similarly, just for consistency with the other examples?

```suggestion
using Pulumi;

var config = new Config();
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+var region = config.Require("region");
+```
+
+{{% /choosable %}}
+
+{{% /chooser %}}
+
+This code will reach from the `Pulumi.<stack-name>.yaml` file and give you access to the configuration values for the prefix you want to load.
+
+So if we want project level config that works across all stacks, we can "hard code" the configuration. It sounds too simple, right?
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:76]:
```suggestion
So if you want project level config that works across all stacks, you can "hard code" the configuration. It sounds too simple, right?
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+{{< chooser language "typescript,go,python,csharp" >}}
+
+{{% choosable language typescript %}}
+
+```typescript
+const projectLevelConfig = {
+  region: "us-west-2",
+  encryptionKmsKey: "arn:aws:kms:us-west-2:...",
+  issueEmail: "bugs@ellingsonmineral.com"
+}
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:87]:
```suggestion
};
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+  { "encryptionKmsKey", "arn:aws:kms:us-west-2:..." },
+  { "issueEmail", "bugs@ellingsonmineral.com"
+}
+```
+
+{{% /choosable %}}
+
+{{% /chooser %}}
+
+These objects/structs/dictionaries can all be consumed and imported across the various components and files within your Pulumi program, but they cannot be used at an organizational level across multiple Pulumi program's ... unless you've got a mono-repo. Do you? 😅
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:137]:
```suggestion
These objects/structs/dictionaries can all be consumed and imported across the various components and files within your Pulumi program, but they cannot be used at an organizational level across multiple Pulumi programs ... unless you've got a mono-repo. Do you? 😅
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+## Organization Level Config
+
+So if we want to provide organizational level config, we need to fall deeper into the programming language eco-system that we're using to build our infrastructure. Every programming language has a way to parse and consume JSON and YAML files. Better yet, every programming language has a way to fetch a remote JSON/YAML file and give you a object/struct/dictionary.
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:141]:
```suggestion
So if you want to provide organizational level config, you need to fall deeper into the programming language ecosystem that you're using to build your infrastructure. Every programming language has a way to parse and consume JSON and YAML files. Better yet, every programming language has a way to fetch a remote JSON/YAML file and give you a object/struct/dictionary.
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+So really, there's no reason you can't use a remote JSON file to provide a common configuration object across all your Pulumi programs. That being said, there are definitely reason's you **shouldn't** use a remote JSON file. Maintaining a remote JSON file is much more of a burden than it will initially seem. As you add more values to your JSON file and more teams/projects begin to rely on it, you'll start to feel the pain of schema management. How do I know that the JSON I'm pulling down has the values I need and that fields haven't been changed or replaced? 😬
+
+So while I 💯 feel like you shouldn't do this, you __can__ if you really need to. Just make sure you understand the tradeoffs and enforce a schema.
+
+A common way to manage schema is to use a [JSON Schema](https://json-schema.org/) or [CUE](https://cuelang.org) to define the structure of your JSON file.
+
+By using one of these methods, you can publish a schema that is available within your organization and people can have confidence that the value they pull remotely can be deserialized to a strict type. Using CI/CD you can also ensure the value itself confirms to the schema before updating the public document.
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:149]:
```suggestion
By using one of these methods, you can publish a schema that is available within your organization and people can have confidence that the value they pull remotely can be deserialized to a strict type. Using CI/CD you can also ensure the value itself conforms to the schema before updating the public document.
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+    },
+    "issueEmail": {
+      "type": "string",
+      "default": "bugs@ellingsonmineral.com"
+    }
+  }
+}
+```
+
+With CUE, this is much more concise. CUE is really awesome and I encourage you to check it out.
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:173]: Nice, TIL -- I didn't know this was a thing.

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+However, as previously mentioned, the organization level config is prone and ripe with pain, confusion, and delusion. It requires non-trivial tooling and process to ensure that a global document is valid and consistent with downstream consumers.
+
+What I've shown you is a "if you must" approach that I would use myself. We've also not even handled the "merge" of organization level with project level overrides, and then stack level overrides. Of course, with code it's all possible, but it's more and more code that you need to write that we don't want you to need to write, right? Good.
+
+Pulumi wants to make this better for you. While this feature doesn't exist yet, we are actively working on it. If you want to follow along with the progress, I suggest you watch/subscribe to [this issue](https://github.com/pulumi/pulumi/issues/2307).
+
+If you want to share your opinions on how **YOU** would like this feature to be implemented, jump into the comments. We'd love to hear from you!
+
+Speak soon
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:194]:
```suggestion
Speak soon.
```

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 12]
```diff
@@ -0,0 +1,194 @@
+---
+title: "Hierarchical Config: The Interim Solution"
+date: 2022-03-10T10:00:00Z
+draft: false
+meta_desc: In this article, we deploy k3s and use the Command package to retrieve our kubeconfig from the virtual-machine and create a Kubernetes provider
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:5]: This'll need an update.

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md [hunk 13]
```diff
@@ -0,0 +1,194 @@
...
+meta_image: meta.png
+authors: ["david-flanagan"]
+tags: ["Config"]
```
Comment (cnunciato) [inline themes/default/content/blog/2022-03-10-hierarchical-config/index.md:8]:
```suggestion
tags: ["config"]
```

**Review by cnunciato — APPROVED** [review]
Nice! Good tips in here. Mostly lil' nits, apologies if anything seems obnoxious. 😄 Approving to leave them all to your discretion (except the meta-description one -- we should fix that up).

It might also be nice to see a quick example of consuming that project-level config in a short program, since you've gone to the trouble of including some basic config and the configs themselves. Would cap things off a nicely, I think. But up to you.

File: themes/default/content/blog/2022-03-10-hierarchical-config/index.md (hunk 9 above)
Comment (nimbinatus) [reply themes/default/content/blog/2022-03-10-hierarchical-config/index.md:149]: Also, more commas :D

```suggestion
By using one of these methods, you can publish a schema that is available within your organization, and people can have confidence that the value they pull remotely can be deserialized to a strict type. Using CI/CD, you can also ensure the value itself confirms to the schema before updating the public document.
```

**Review by nimbinatus — APPROVED** [review]
Love it! A couple tiny nits, but nothing blocking. Ship it!

**Pull request merged**